src/box/box.cc

```diff
@@ -793,6 +799,172 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
 	return -1;
 }
 
+/* The mempool for allocating rows. */
+static struct mempool wal_row_pool;
+
+/* The lsregion for allocating row bodies. */
+static struct lsregion wal_row_lsr;
+
+/* Storage, adding rows to a rlist to separate mixed transactions. */
+struct wal_row {
+	/* Base class. */
+	struct xrow_header row;
+	/* List of rows to stacked in the nodes_rows array. */
+	struct rlist in_row_list;
+	/* A growing identifier to track lsregion allocations. */
+	int64_t lsr_id;
+};
+
+static void
+wal_mem_init(void)
+{
+	mempool_create(&wal_row_pool, &cord()->slabc,
+		       sizeof(struct wal_row));
+	lsregion_create(&wal_row_lsr, &runtime);
+}
+
+static void
+wal_mem_destroy(void)
+{
+	lsregion_destroy(&wal_row_lsr);
+	mempool_destroy(&wal_row_pool);
+}
+
+/*
+ * Callback to stash row upon receipt, used to recover mixed transactions.
+ */
+static struct wal_row *
+wal_stream_save_row(struct xrow_header *row)
+{
+	struct wal_row *new_row =
+	(struct wal_row *)mempool_alloc(&wal_row_pool);
+	if (new_row == NULL) {
+		tnt_raise(OutOfMemory, sizeof(*new_row), "mempool_alloc",
+			  "xrow");
+	}
+	memcpy(new_row, (struct wal_row *)row, sizeof(struct wal_row));
+	return new_row;
+}
+
+/*
+ * Callback to stash row bodies upon receipt, used to recover mixed
+ * transactions.
+ */
+static void
+wal_stream_save_body(struct wal_row *row)
```

Comment:
Done.

Reply:
Thanks!
In `applier.cc` `alloc_row` and `save_body` are separated because `alloc_row` creates a placeholder for a row to be read into by `coio_read_xrow`. And only after doing that you can actually save the body.

In our case now there is no need to separate the functions because you already receive a ready row and need to copy it completely at once.